### PR TITLE
MODINREACH-166 Check out to borrowing site for Item Hold Transaction

### PR DESCRIPTION
**Pull request by oliinyko**
## Purpose
Provide an API endpoint to process checking out an item from the owning site to the borrowing site for an Item Hold transaction.
https://issues.folio.org/browse/MODINREACH-166

## Approach
- Add endpoint to handle check out for Item Hold
- Add a service for transaction flow actions
- Implement check out to borrowing site for Item Hold
- Refactoring patron hold receive item action flow
- Add tests

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -373,6 +373,14 @@
           "modulePermissions": [
             "circulation.check-in-by-barcode.post"
           ]
+        },
+        {
+          "methods": ["PUT"],
+          "pathPattern": "/inn-reach/transactions/{itemBarcode}/check-out-item/{servicePointId}",
+          "permissionsRequired": ["inn-reach.inn-reach-transaction.item.put"],
```
Comment (dmtkachenko) [inline descriptors/ModuleDescriptor-template.json:395]: I see that the same permission is used for another endpoint:
> "methods": ["PUT"],
          "pathPattern": "/inn-reach/transactions/{id}/receive-item/{servicePointId}",
          "permissionsRequired": ["inn-reach.inn-reach-transaction.item.put"],
          "modulePermissions": [
            "circulation.check-in-by-barcode.post"
          ]

Has it been done on purpose?

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  }
+
+  private void reportItemReceived(InnReachTransaction transaction) {
+    var centralCode = transaction.getCentralServerCode();
+    var requestPath = resolveItemReceivedPath(transaction.getTrackingId(), centralCode);
+
+    try {
+      innReachExternalService.postInnReachApi(centralCode, requestPath);
+    } catch (InnReachException e) {
+      log.warn("Unexpected D2IR response: {}", e.getMessage(), e);
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Have you thought about reporting the issue back to UI? My point is that the whole action would be completed only partially in this case: everything that belongs to Folio is successful, but Inn-Reach call fails. 
With only logging the error, user will be presented with incomplete result, don't you think?

ps: the same goes to `reportItemShipped` method

File: descriptors/ModuleDescriptor-template.json (hunk 1 above)
Comment (oliinyko) [reply descriptors/ModuleDescriptor-template.json:395]: Yes, according to requirements they both should use "permission to modify transactions"

Comment (dmtkachenko) [reply descriptors/ModuleDescriptor-template.json:395]: I'm not sure the requirement is understood correctly. Usually each endpoint has its own permission
Could you double check with PO?

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 2 above)
Comment (oliinyko) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Right now d2ir API doesn't handle these endpoints appropriately. I assume that the operations may not work as the tracking ids that we are sending are not known by central server

The issue was discussed with PO and for now we agreed to suppress the error https://issues.folio.org/browse/MODINREACH-150?focusedCommentId=116734&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-116734

File: descriptors/ModuleDescriptor-template.json (hunk 1 above)
Comment (oliinyko) [reply descriptors/ModuleDescriptor-template.json:395]: Checked

> The same one is fine

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java (hunk 2 above)
Comment (oliinyko) [reply src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: Added a new task to review the flow when d2ir environment is ready for testing these endpoints
https://issues.folio.org/browse/MODINREACH-192

**Review by dmtkachenko — APPROVED** [review]

**Pull request merged**